observer.cpp

```diff
@@ -87,5 +87,38 @@ void test_observer()
         assert(o.get_params() == params);
     }
 #endif
+
+//#define FIX_ISSUE_81
+#ifdef FIX_ISSUE_81
+  ///The observer stores inputs and optimal values
+  {
+    observer o;
+
+    //Give sim some non-default inputs and optimal
+    simulation s{};
+    int n_repeats = 100;
+
+    for(int i = 0; i != n_repeats; ++i){
+        tick(s);
+        assert(o.get_input()[i] != s.get_input());
+        assert(o.get_optimal()[i] != s.get_optimal());
```

Comment:
In this case `observer` does not know for how many ticks it will have to register data, and I suspect that indeed it will not have an `i`th element in its `vector`s on its as they will be initialized empty and in this spot of the code they will still have `size() == i - 1`, so in this case I guess `.back()` is what you want, or probably not even as `.back()` might not work on an `.empty()` vector (maybe put an `if` clause for the first iteration). 
Else you can change the constructor of observer so that it knows in advance the number of elements it will have to store and set the size of its memeber vectors accordingly, but this would require quite a lot of redesign and its own set of issues.
```suggestion
        assert(o.get_input().back() != s.get_input());
        assert(o.get_optimal().back() != s.get_optimal());
```

Reply:
Ok no, definitely put an `if` clause to check the vector is not `empty()`, see [here](https://en.cppreference.com/w/cpp/container/vector/back).